File: simulator/spells/hunger_of_hadar.py
from functools import cache

from cachetools import cached
from cachetools.keys import hashkey

from ..action_resolver import resolve_dmg_saving_throw
from ..actions.action_types import BonusAction
from ..battle_map import Map, map_position_toggled_cache
from ..combatant_coords import Coords
from ..effects.aoe_spheric_effect import AoeSphericEffect
from ..effects.effect import EffectType
from ..effects.limited_duration_effect import LimitedDurationEffect
from ..spells.spell import SpellStats
from ..misc import SavingThrow, DamageType, avg_roll, roll_spell_dmg
from ..conditions import Conditions, Condition, is_affected_by_any, get_swallower, apply_condition, \
    remove_condition
from ..actions.actoid import Actoid, ActoidFlags, FactoryFlags
from ..threat_utils import mean_dmg_dc_attack
from ..threat_interfaces import DirectThreat, AoEThreat
from ..factory_interfaces import DirectThreatFactory
import numpy as np


class HungerOfHadarFactory(DirectThreatFactory):
    level = 3
    range = SpellStats.Range.FEET_150.value
    target = SpellStats.Target.RADIUS_20
    duration = SpellStats.Duration.INSTANTANEOUS
    concentration = True
    type = SpellStats.Type.HARMFUL

    def __init__(self, dc, action_type, caster, resource):
        super().__init__()
        self.flags |= FactoryFlags.DEX_SAVE_APPLIES
        self.dc = dc
        self.action_type = action_type  # HUNGER_OF_HADAR, QUICKENED_HUNGER_OF_HADAR
        self.saving_throw = SavingThrow.DEX
        self.dmg_dice = "2d6"
        self.combatant = caster
        self.resource = resource
        self.dmg_type = DamageType.Cold

    def __str__(self):
        """
        Important for FSM building
        """
        return "HungerOfHadarFactory"

    def get_ability_name(self):
        return "Hunger of Hadar"

    def find_best_args(self, combatant):
        # TODO Deprecated
        battle_map = Map.get()
        coord, _ = battle_map.find_best_placement_harmful_circular(combatant, HungerOfHadarFactory.range, SpellStats.TRANSLATE_RADIUS[HungerOfHadarFactory.target], self)
        return coord

    def create_all(self, previous_action_in_dag=None):
        # Here there really is no need to iterate over all coords. Just find the best score
        return [HungerOfHadar(self.find_best_args(self.combatant), self)]

    def create(self, coord):
        return HungerOfHadar(coord, self)

    def calculate_max_threat(self):
        return HungerOfHadar(self.find_best_args(self.combatant), self).calculate_threat()

    def calculate_threat_to_target(self, target, **kwargs):
        """
        Calculates threat to one specific target
        """
        # The 0.5 is a heuristic which expresses the fact that most targets would leave the area immediately
        mean_dmg = min(target.curr_hp, mean_dmg_dc_attack(self.dc, self.dmg_dice, False, target.saving_throws[self.saving_throw], target.is_resistant_to(DamageType.Acid)))
        return avg_roll(self.dmg_dice) + 0.5 * mean_dmg

    def calculate_threat_to_target_delta(self, target, modifiers, *args, **kwargs):
        """
        Calculates the threat delta of the factory to a specific target given stat modifications
        """
        return 0 # No need


class HungerOfHadar(Actoid, LimitedDurationEffect, AoeSphericEffect, DirectThreat, AoEThreat):

    def __init__(self, coord, factory,  **kwargs):
        Actoid.__init__(self, ActoidFlags.IS_SPELL)
        LimitedDurationEffect.__init__(self, factory.combatant, turns=10)
        AoeSphericEffect.__init__(self, factory.combatant, coord, SpellStats.TRANSLATE_RADIUS[HungerOfHadarFactory.target])
        self.factory = factory

    def __str__(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_HUNGER_OF_HADAR else "") + f"Hunger Of Hadar at {np.squeeze(self.origin)}"

    def get_effect_type(self):
        return EffectType.HUNGER_OF_HADAR

    def shorthand_str(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_HUNGER_OF_HADAR else "") + "Hunger Of Hadar"

    def on_start_of_turn(self, combatant):
        apply_condition(combatant, Condition(Conditions.BLINDED, self.factory.combatant, self))
        dmg = roll_spell_dmg(self.factory.dmg_dice)
        combatant.receive_dmg(dmg, self.factory.dmg_type)
        Map.get().remove_combatant_if_dead(combatant)

    def on_end_of_turn(self, combatant):
        apply_condition(combatant, Condition(Conditions.BLINDED, self.factory.combatant, self))
        dmg = roll_spell_dmg(self.factory.dmg_dice)
        self.factory.dmg_type = DamageType.Acid
        resolve_dmg_saving_throw(self, dmg, combatant, False, True)
        self.factory.dmg_type = DamageType.Cold

    def on_enter(self, combatant):
        apply_condition(combatant, Condition(Conditions.BLINDED, self.factory.combatant, self))

    def on_move_within(self, combatant):
        pass

    def on_exit(self, combatant):
        remove_condition(combatant, Conditions.BLINDED, self.factory.combatant)

    # def is_affecting(self, combatant):
    #     battle_map = Map.get()
    #     coords = self.get_affected_coords()
    #     return battle_map.get_hop_distance_coords(battle_map.get_combatant_position(combatant).get(), coords) == 0

    def activate(self, **kwargs):
        Map.get().effect_tracker.add(self)
        self.factory.combatant.concentration_effect = self
        # TODO make the area difficult terrain

    def deactivate(self):
        # TODO remove difficult terrain
        self.factory.combatant.break_concentration()

    def deactivate_for_combatant(self, combatant):
        assert False

    @map_position_toggled_cache
    def calculate_threat(self, **kwargs):
        battle_map = Map.get()
        affected = battle_map.get_combatants_affected_by_sphere_aoe(self.factory.combatant, HungerOfHadarFactory.target, HungerOfHadarFactory.type, self.origin)
        acc = 0
        for aff in affected:
            acc += avg_roll(self.factory.dmg_dice)  # the initial cold dmg
            # The 0.5 is a heuristic which expresses the fact that most targets would leave the area immediately
            mean_dmg = min(aff.curr_hp, mean_dmg_dc_attack(self.factory.dc, self.factory.dmg_dice, False, aff.saving_throws[self.factory.saving_throw], aff.is_resistant_to(DamageType.Acid)))
            acc += 0.5 * mean_dmg
            acc *= (1 if battle_map.teams.are_enemies(self.factory.combatant, aff) else -3)
        return acc

    def clear_cache(self):
        self.calculate_threat.cache_clear()
        #self.get_eligible_coords.cache_clear()

    def calculate_threat_delta(self, modifiers, *args, **kwargs):
        return 0  # Not relevant for this ability

    def threat_on_end_of_turn(self, target, *args, **kwargs):
        return min(target.curr_hp, mean_dmg_dc_attack(self.factory.dc, self.factory.dmg_dice, False, target.saving_throws[self.factory.saving_throw], target.is_resistant_to(DamageType.Acid)))

    def threat_on_enter(self, target, *args, **kwargs):
        return 0

    def threat_on_start_of_turn(self, target, *args, **kwargs):
        threat = avg_roll(self.factory.dmg_dice)
        return threat if not target.is_resistant_to(self.dmg_type) else threat / 2

    def threat_on_move_within(self, target, *args, **kwargs):
        return 0

    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        if get_swallower(self.factory.combatant):
            return None
        battle_map = Map.get()
        if not is_affected_by_any(self.factory.combatant, Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
            return battle_map.get_free_coords_in_cartesian_range(Coords(self.origin),  # not actually combatant coords
                                                                 distances,
                                                                 inflate_to_dist=self.factory.combatant.size.value,
                                                                 rng=HungerOfHadarFactory.range, combatant=self.factory.combatant)
        elif battle_map.get_cartesian_distance_coords(battle_map.get_combatant_position(self.factory.combatant).get(), np.array([self.origin])) <= HungerOfHadarFactory.range:
            return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]
        return None

File: c++/spells/hunger_of_hadar.cpp
// hunger_of_hadar.cpp
#include "spells/hunger_of_hadar.hpp"
#include "core/battle_map.hpp"
#include "core/teams.hpp"
#include "core/combatant.hpp"
#include "effects/effect_tracker.hpp"
#include "effects/aoe_effect.hpp"
#include "core/conditions.hpp"
#include "core/action_resolver.hpp"
#include "core/geometry.hpp"

namespace enc
{
  HungerOfHadarFactory::HungerOfHadarFactory(int dc, AbilityType abilityType, Combatant *caster, Resource *resource)
      : DirectThreatFactory("HungerOfHadarFactory", "Hunger of Hadar", caster, abilityType), _dc(dc), _abilityType(abilityType), _resource(resource),
        _savingThrow(SavingThrow::DEX), _dmgDice({{2, 6}})
  {
    setFlag(FactoryFlags::DEX_SAVE_APPLIES);
  }

  std::vector<std::shared_ptr<Actoid>> HungerOfHadarFactory::createAll(void *previousActionInDag)
  {
    auto &battleMap = BattleMap::getInstance();
    auto [coord, _1, _2] = battleMap.findBestPlacementHarmfulCircular(_combatant, static_cast<int>(HungerOfHadarFactory::range),
                                                                    TRANSLATE_RADIUS.at(HungerOfHadarFactory::target));
    return {std::make_shared<HungerOfHadar>(coord, *this)};
  }

  std::shared_ptr<Actoid> HungerOfHadarFactory::create(void *target) { return std::make_shared<HungerOfHadar>(*static_cast<Coord *>(target), *this); }

  double HungerOfHadarFactory::calculateThreatToTarget(Combatant *target, const Kwargs &kwargs) const
  {
    // The 0.5 is a heuristic expressing that most targets would leave the area
    auto avgDmg = std::min(static_cast<double>(target->getCurrentHp()),
                           meanDmgDcAttack(_dc, {_dmgDice}, false, target->getSavingThrow(_savingThrow), target->isImmuneTo(DamageType::Acid),
                                           target->isResistantTo(DamageType::Acid)));

    return avgRoll(_dmgDice) + 0.5 * avgDmg;
  }

  double HungerOfHadarFactory::calculateThreatToTargetDelta(Combatant *target, const ThreatModifiers &modifiers) const
  {
    return 0.0; // No need
  }

  double HungerOfHadarFactory::calculateMaxThreat() const
  {
    auto &battleMap = BattleMap::getInstance();
    auto [coord, _1, _2] = battleMap.findBestPlacementHarmfulCircular(_combatant, static_cast<int>(HungerOfHadarFactory::range),
                                                                    TRANSLATE_RADIUS.at(HungerOfHadarFactory::target));
    HungerOfHadar effect(coord, *this);
    return effect.calculateThreat({});
  }

  HungerOfHadar::HungerOfHadar(const Coord &coord, const HungerOfHadarFactory &factory)
      : Effect(factory._combatant), AoeEffect(factory._combatant),
        Actoid(const_cast<HungerOfHadarFactory &>(factory), ActoidFlags::IS_SPELL, factory._abilityType),
        LimitedDurationEffect(factory._combatant, 10), AoeSphericEffect(factory._combatant, coord, TRANSLATE_RADIUS.at(HungerOfHadarFactory::target)),
        _coord(coord), _factory(factory)
  {}

  void HungerOfHadar::onStartOfTurn(Combatant *combatant)
  {
    combatant->applyCondition(std::make_shared<Condition>(Conditions::BLINDED, _factory._combatant, this));
    int damage = rollDice(_factory._dmgDice);
    combatant->receiveDmg(damage, _factory.dmgType);
    BattleMap::getInstance().removeCombatantIfDead(*combatant);
  }

  void HungerOfHadar::onEndOfTurn(Combatant *combatant)
  {
    combatant->applyCondition(std::make_shared<Condition>(Conditions::BLINDED, _factory._combatant, this));
    int damage = rollDice(_factory._dmgDice);

    // Temporarily change damage type for acid damage
    auto originalDmgType = _factory.dmgType;
    resolveDmgSavingThrow(_factory._savingThrow, _factory._dc, shorthandStr(), damage, DamageType::Acid, combatant, false, true);
  }

  void HungerOfHadar::onEnter(Combatant *combatant)
  {
    combatant->applyCondition(std::make_shared<Condition>(Conditions::BLINDED, _factory._combatant, this));
  }

  void HungerOfHadar::onMoveWithin(Combatant *combatant)
  {
    // No effect when moving within area
  }

  void HungerOfHadar::onExit(Combatant *combatant) { combatant->removeCondition(Conditions::BLINDED, _factory._combatant); }

  void HungerOfHadar::activate(const Kwargs &kwargs)
  {
    auto &effectTracker = EffectTracker::getInstance();
    effectTracker.add(Effect::shared_from_this());
    _factory._combatant->setConcentrationEffect(Effect::shared_from_this());
    // TODO: Make area difficult terrain
  }

  void HungerOfHadar::deactivate()
  {
    _factory._combatant->breakConcentration();
    // TODO: Remove difficult terrain
  }

  bool HungerOfHadar::deactivateForCombatant(Combatant *combatant) { throw std::runtime_error("Not implemented"); }

  double HungerOfHadar::calculateThreat(const Kwargs &kwargs)
  {
    auto &battleMap = BattleMap::getInstance();
    Teams &teams = Teams::getInstance();
    auto affected = battleMap.getCombatantsAffectedBySphereAoE(_factory._combatant, HungerOfHadarFactory::target, HungerOfHadarFactory::type, _coord);

    double totalThreat = 0.0;
    for(auto *target : affected)
      {
        // Initial cold damage
        totalThreat += avgRoll(_factory._dmgDice);

        // Potential acid damage with 0.5 multiplier (assuming targets try to leave)
        double avgDmg = std::min(static_cast<double>(target->getCurrentHp()),
                                 meanDmgDcAttack(_factory._dc, {_factory._dmgDice}, false, target->getSavingThrow(_factory._savingThrow),
                                                 target->isImmuneTo(DamageType::Acid), target->isResistantTo(DamageType::Acid)));
        totalThreat += 0.5 * avgDmg;

        // Adjust for friendly fire
        totalThreat *= (teams.areEnemies(*_factory._combatant, *target) ? 1.0 : -3.0);
      }
    return totalThreat;
  }

  double HungerOfHadar::calculateThreatDelta(const ThreatModifiers &modifiers) const
  {
    return 0.0; // Not relevant for this ability
  }

  std::string HungerOfHadar::toString() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_HUNGER_OF_HADAR) ? "Quickened " : "";
    std::stringstream ss;
    ss << _coord;
    return prefix + "Hunger of Hadar at " + ss.str();
  }

  std::string HungerOfHadar::shorthandStr() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_HUNGER_OF_HADAR) ? "Quickened " : "";
    return prefix + "Hunger of Hadar";
  }

  EffectType HungerOfHadar::getEffectType() const { return EffectType::HUNGER_OF_HADAR; }

  const CoordVector &HungerOfHadar::getAffectedCoords() const { return SphericAoe::getAffectedCoords(); }

  std::optional<CoordVector>
  HungerOfHadar::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    if(_factory._combatant->getSwallower())
      {
        return std::nullopt;
      }

    auto &battleMap = BattleMap::getInstance();

    if(!_factory._combatant->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        return battleMap.getFreeCoordsInCartesianRange(Coords(_origin), distances, _factory._combatant->getSize(),
                                                       static_cast<int>(HungerOfHadarFactory::range), _factory._combatant->_instanceId);
      }

    const Coords &combatantPos = battleMap.getCombatantCoordinates(*_factory._combatant);
    if(getCartesianDistanceCoords(combatantPos, _origin) <= static_cast<double>(HungerOfHadarFactory::range))
      {
        return CoordVector{combatantPos.getRoot()};
      }

    return std::nullopt;
  }

  size_t HungerOfHadar::hash() const
  {
    size_t h = std::hash<int>{}(static_cast<int>(getAbilityType()));
    h ^= std::hash<int>{}(static_cast<int>(getFlags())) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<int>{}(_coord[0]) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<int>{}(_coord[1]) + 0x9e3779b9 + (h << 6) + (h >> 2);
    return h;
  }

  bool HungerOfHadar::equals(const Actoid &other) const
  {
    if(auto *hungerOfHadar = dynamic_cast<const HungerOfHadar *>(&other))
      {
        return getAbilityType() == other.getAbilityType() && getFlags() == other.getFlags() && _coord == hungerOfHadar->_coord;
      }
    return false;
  }
}
